From patchwork Fri Apr 23 15:58:24 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Ashish Kalra <Ashish.Kalra@amd.com>
X-Patchwork-Id: 12220819
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.7 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,URIBL_BLOCKED,USER_AGENT_GIT autolearn=ham autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 62013C433ED
	for <kvm@archiver.kernel.org>; Fri, 23 Apr 2021 15:58:41 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 2FAA860C40
	for <kvm@archiver.kernel.org>; Fri, 23 Apr 2021 15:58:41 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S230294AbhDWP7P (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Fri, 23 Apr 2021 11:59:15 -0400
Received: from mail-bn8nam12on2089.outbound.protection.outlook.com
 ([40.107.237.89]:21216
        "EHLO NAM12-BN8-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S243068AbhDWP7O (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 23 Apr 2021 11:59:14 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=ksJe6jpa9G4Wn17zN63uA/b78ftNWToON2m7Z5eUCO6uaGfpeuoU6NfAYVxeGGtTbVlNgsaLz/PTwNzEYSL5idjO1NBFW8oASpzAe1GJn0mcpHRTguyIdNzVmVe0qhEKqvwG3Ik+MaKsMU+tFVlFcD5bUCguiAOuIGb1DV5yrUKA53LG4LQyYWaAg1MHhg5OplxLuafbMdqnjGIw1HkWyOBgUJfPH6KmipfMqMvokQ8dK1vktS0et/wSNKocUaAXkWHyMQLUus5qYyC68NswxtTZdb/uAoaWsxqICbMg0qMSbDAKF/kRGDpgNAHk2ukyQg+Y89tfSHdW7tGKYm7yow==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=kbX6VSM3gY2hDVfwBpZ4zpupPxEYMewOcFc83mr9Buc=;
 b=R9lmtPX7MT0btvU3RwwN/U5raVb3P9EYLI6jSeh0VOxFUfVldHZ80rn6fYoK8CXUqG1b9BwA7YEJHfg5WuGVsLEvItebsISgKv61yrlxsPa9n9VEYbkiLU0SmZGR7W1cHx5BCvORvx94epmmVBk6k7w/zhfm9k8vYlbrH7vgLpMZ5U+Mn3d+e1d6vVSIqcGqDQlCZYF097BM2kJRsrYbM737NmzfKyTDkN6JVw+pA90xFEmVF7H8JHwtvpOC1api2YUaXNKKb+tpkyXRb7jX3vGorHKqlaMjOy72HBLi+192yvnEP1kBkWYib9zt8MSVXISKLfOFIBCEY1IH2nl2CQ==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=amd.com; dmarc=pass action=none header.from=amd.com; dkim=pass
 header.d=amd.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=amd.com; s=selector1;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=kbX6VSM3gY2hDVfwBpZ4zpupPxEYMewOcFc83mr9Buc=;
 b=VOeRUdY+jJpQmXKGXaWCdUm42qZ3WD/44A8HCSxtyMjNiEjQgogznL13JgxaGevG/qmoumqTxxchMiHQRF6IPaKE8oVujVHhcO3jQedjteZHzr1kScJUVgPOUe5eWHo3FlabllPnCLfbb8eesQyCeTHBk9xEsWOlNB1kVHA71ps=
Authentication-Results: redhat.com; dkim=none (message not signed)
 header.d=none;redhat.com; dmarc=none action=none header.from=amd.com;
Received: from SN6PR12MB2767.namprd12.prod.outlook.com (2603:10b6:805:75::23)
 by SN6PR12MB2639.namprd12.prod.outlook.com (2603:10b6:805:75::25) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4042.18; Fri, 23 Apr
 2021 15:58:34 +0000
Received: from SN6PR12MB2767.namprd12.prod.outlook.com
 ([fe80::1fb:7d59:2c24:615e]) by SN6PR12MB2767.namprd12.prod.outlook.com
 ([fe80::1fb:7d59:2c24:615e%7]) with mapi id 15.20.4065.025; Fri, 23 Apr 2021
 15:58:34 +0000
From: Ashish Kalra <Ashish.Kalra@amd.com>
To: pbonzini@redhat.com
Cc: tglx@linutronix.de, mingo@redhat.com, hpa@zytor.com,
        joro@8bytes.org, bp@suse.de, thomas.lendacky@amd.com,
        x86@kernel.org, kvm@vger.kernel.org, linux-kernel@vger.kernel.org,
        srutherford@google.com, seanjc@google.com,
        venu.busireddy@oracle.com, brijesh.singh@amd.com
Subject: [PATCH v2 1/4] KVM: x86: invert KVM_HYPERCALL to default to VMMCALL
Date: Fri, 23 Apr 2021 15:58:24 +0000
Message-Id: 
 <76ad1a3f7ce817e8d269a6d58293fc128678affc.1619193043.git.ashish.kalra@amd.com>
X-Mailer: git-send-email 2.17.1
In-Reply-To: <cover.1619193043.git.ashish.kalra@amd.com>
References: <cover.1619193043.git.ashish.kalra@amd.com>
X-Originating-IP: [165.204.77.1]
X-ClientProxiedBy: SA0PR11CA0078.namprd11.prod.outlook.com
 (2603:10b6:806:d2::23) To SN6PR12MB2767.namprd12.prod.outlook.com
 (2603:10b6:805:75::23)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ashkalra_ubuntu_server.amd.com (165.204.77.1) by
 SA0PR11CA0078.namprd11.prod.outlook.com (2603:10b6:806:d2::23) with Microsoft
 SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4065.21 via Frontend
 Transport; Fri, 23 Apr 2021 15:58:34 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: 5d28f07b-add5-4ab9-f28b-08d90670a639
X-MS-TrafficTypeDiagnostic: SN6PR12MB2639:
X-MS-Exchange-Transport-Forked: True
X-Microsoft-Antispam-PRVS: 
 <SN6PR12MB2639E3F467017B9E309AFA848E459@SN6PR12MB2639.namprd12.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:5516;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 mkziyzWRth5e/CaMoZEhdZUXWWMzGWDCvc2cm3unEbW4nZnKKAbicBArySox2X/nfCGUEBFZYVR5KWCBlmsQJNJpf/bWyQxMQ/83kAMfc81wSDNWuq60av7hgg+zDV0qD0Kj7lSZ1dY7Lr5VkopMODhY1ew8Lk7S40wN/Z1GjF5fEadQqus8GxgRoARtpHSLTMr/0AdMQTwUtXadZ1OBzFRJ6yo+9lCW7naByqa5D/aq2lh3wREdXACxOQv/JFYnvJznJ9Oz35GczL26FeqVq2HPuZOS2oeztx5ywQi4L2VD8L5Y4cHhuii+dgzOPCmQ6ghtSEcwb0BRYVlCReO1RYNoZaKKnL974tAXM/KT/gZsH3KUDuvTQ+mBKpdp/w9ubYWeve1EG5XeJhg6Ucf4+8jF9uAFc/ZJ8QJtWeLCF4sEi9smoZV/bbpPV/ELxWiSkSbu6+/huWfAYZNPHEpAYKx58RwWM8vJGltm3zdYV+CwINjeK+YKI0lkbi2bIWozsKbyuDUicMlFDaIsatxnyyatnh2iezRcwuGK9h1Ql9HVTG48Aq2ZmlX+kJ8VCYR/+YNOP4Mn5C6afVybWD9Mib0mLLf7jwEcsjFdszqbSrC15pmvuRwebWZwhgoQVpMXRwPGptvcSbQOsFDZaAXANsrG5JfwYBvNR0MDE0Sl77A=
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR12MB2767.namprd12.prod.outlook.com;PTR:;CAT:NONE;SFS:(4636009)(366004)(136003)(346002)(396003)(376002)(39850400004)(66556008)(66946007)(6916009)(4326008)(478600001)(956004)(8936002)(26005)(186003)(316002)(6666004)(7416002)(52116002)(2616005)(6486002)(8676002)(5660300002)(2906002)(38100700002)(83380400001)(38350700002)(36756003)(66476007)(16526019)(86362001)(7696005);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 62ZnAa/BMWbdQBtPHfdu3bve5+E3xVbAqTIWScI1++cZ6bVja530mGioQuMR2bSZ6uTpXYr8CI9utUb79O5CkQpKiKrcP7yIuAAxzfse5qLb5OP5F2DzHvkb/5mFj73RlgHBdCTTa7vWPqFpwJlaqVUkE2DgsO1F0r7QxKt3gETE54q5pu2U/xoue9Et9rWqpn0qPL4QZ4pLhfgwJQS6vu25ecb8leHTDlaVtpoYwkiU3WD7/7FgEmFrPBMLJxhSd9C7GEbzxg/fPrh2f1xP87u3k0a8F+WqTex2I33kZaMqR1gn6BDEq/gl6vKvDF4rC3OQqSPfg5rh8iwlJOhFoOwZJR+/cdzL6DEbjLSo81ZpLvoMeTWPS9CJv4tOD5no+9DkoLs5/pv5YuzY6xBAhwysNBuxziT0CjJzhAIvaup3g2BJjosOscP2/p6E8+SkKwsGoiAMsqReAE7+r3YS2S+v4PQqu/HnpC5aqmt+Y2p29fF6qtv2UuUFdXVNlpxoPjqFplbmGkYNtSG2nkdGSQqGuV+D9GBEkYqMZ0lFYayRYP/HfaUVTtCP1daqSw5NDQoOohkFubFA2r31rKhQeo8EV+0IoPXjzJq9PmEiM2hgtBNqz6H7QAnueurbXKr75P13KWPFadmho50xXwk/Yh31gP2Ebvw1KccVgfJNMFz6NR6HzNAd0qLoN/ESg9ZqrTuFTodfGEv89wPe+QLB4OP37cRwiUl2QJJ7ncqINrbjFgSQbPcVKXQldUzKDlCiRh9z/uSb4GKzP7FXmF9wzf7XnA6lgHK2dhAfx7qgYG8te2n+VIg6l8MhWHpxr1jf9z9X6IbHZCRNmuFQDEzQ86rVMb4NIHOnhkbXXM6FVD8mhgIa8Fci/jckvirdi1jpVilpnR3B7u+AUdxB0gV88hG2CfCbPTec5zaGHdCHm/Pe8yU1uE2XwZryLHgfAF5Yx/DoPkok7ypTdSX3uoKlV2uXsU8d3NYyo7s4Km94DZeXAb33azrHLPcX03ZbOAFj4uw8xEJo8JbEtUytqLf2o9GAA2OX44Ao/RsSEmXLJDT1tNs1l91y/VwlQn5mo5tcVxsydinKmEqv45Gt0DsCX2U72rTqUxGyP3120XimkgRtKIcgKxM6rtwImZPjcWN+r0lTIJVuILwb9lha8EG3QFYHGzDX2W9L5cOC9Zsyq/AxDMwMU9/bsQByIE+9u/naqeCwXcb0UwrFmUdTz8EViFmOfPUggjrSYUltwv2+w3RnVRYIlCq32SF9dyRILR9sdZGShfdGKmsRfeXjLp52sgQ2JDNTPtdXDZuT2Q9N5p5XATvYRDM8H3c1HkdKIIMY
X-OriginatorOrg: amd.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 5d28f07b-add5-4ab9-f28b-08d90670a639
X-MS-Exchange-CrossTenant-AuthSource: SN6PR12MB2767.namprd12.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 23 Apr 2021 15:58:34.8612
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 3dd8961f-e488-4e60-8e11-a82d994e183d
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 hNe+5IaySuLt9JjE2WuQtsAPe894mu3G6FPi8LTZZMYX+VA0gZfaJbVWExcifJM/CJEFAJu6+GygavBrTyNkKQ==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SN6PR12MB2639
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Ashish Kalra <ashish.kalra@amd.com>

KVM hypercall framework relies on alternative framework to patch the
VMCALL -> VMMCALL on AMD platform. If a hypercall is made before
apply_alternative() is called then it defaults to VMCALL. The approach
works fine on non SEV guest. A VMCALL would causes #UD, and hypervisor
will be able to decode the instruction and do the right things. But
when SEV is active, guest memory is encrypted with guest key and
hypervisor will not be able to decode the instruction bytes.

So invert KVM_HYPERCALL and X86_FEATURE_VMMCALL to default to VMMCALL
and opt into VMCALL.

Cc: Thomas Gleixner <tglx@linutronix.de>
Cc: Ingo Molnar <mingo@redhat.com>
Cc: "H. Peter Anvin" <hpa@zytor.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Joerg Roedel <joro@8bytes.org>
Cc: Borislav Petkov <bp@suse.de>
Cc: Tom Lendacky <thomas.lendacky@amd.com>
Cc: x86@kernel.org
Cc: kvm@vger.kernel.org
Cc: linux-kernel@vger.kernel.org
Signed-off-by: Brijesh Singh <brijesh.singh@amd.com>
Signed-off-by: Ashish Kalra <ashish.kalra@amd.com>
---
 arch/x86/include/asm/kvm_para.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/include/asm/kvm_para.h b/arch/x86/include/asm/kvm_para.h
index 338119852512..fda2fe0d1b10 100644
--- a/arch/x86/include/asm/kvm_para.h
+++ b/arch/x86/include/asm/kvm_para.h
@@ -19,7 +19,7 @@ static inline bool kvm_check_and_clear_guest_paused(void)
 #endif /* CONFIG_KVM_GUEST */
 
 #define KVM_HYPERCALL \
-        ALTERNATIVE("vmcall", "vmmcall", X86_FEATURE_VMMCALL)
+	ALTERNATIVE("vmmcall", "vmcall", X86_FEATURE_VMCALL)
 
 /* For KVM hypercalls, a three-byte sequence of either the vmcall or the vmmcall
  * instruction.  The hypervisor may replace it with something else but only the

From patchwork Fri Apr 23 15:58:43 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Ashish Kalra <Ashish.Kalra@amd.com>
X-Patchwork-Id: 12220821
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.7 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,URIBL_BLOCKED,USER_AGENT_GIT autolearn=ham autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 0DCDFC433ED
	for <kvm@archiver.kernel.org>; Fri, 23 Apr 2021 15:59:00 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id CDF3D61450
	for <kvm@archiver.kernel.org>; Fri, 23 Apr 2021 15:58:59 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S243109AbhDWP7f (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Fri, 23 Apr 2021 11:59:35 -0400
Received: from mail-eopbgr760081.outbound.protection.outlook.com
 ([40.107.76.81]:1347
        "EHLO NAM02-CY1-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S231858AbhDWP7e (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 23 Apr 2021 11:59:34 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=nMd4JtPRl/Xpt5e9VlMkIGhyD8SYCbZHKsx6T6ku+jsb1yDljoUt/prLPIjq5SkPdLum/ODfeV8a+MwRMxXTTr1iyIuTrP0r4/vcOipANAibFOGIrduUIinpFIMPkOj/sxnNe2PdoepWmY35yvSaQqi+N8g2mmbiDf7pF4cyt7h+mlxpotjPmFzwi4n6aahn4Pw9Ik2OH0EoQjWmFZOFwvMLHB8KX5bMNEC6rAAwUFOBxlAxHPHnl8mD1GG4zL/veS7ufG2Pt11YNIFxRaB9yNkvi0k8ZfkNfnhZOOi1v9ErhoM4jQAPYwuYU3Wsr5tLXAJ0ajPtmMrXAcFmAifDTg==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=V519WsdDsLhQcSitVWrmZPwA4RVtfnxNsUoCkWKP0qw=;
 b=Cer3u7CDgodNogicVkTM6yyxGQHA3OmeupbECK5Y8TETsJkh+YOFEKJU3WSHUIMV97oiVzZhpD8lL/CXHOavHFv00gZVVFSNBEVHWuQ5c7XQWCh1qfG+Y9WDRczgbqfFHKo9BSGvjv7x0I7cj1kFrWgvoD395UqaZDGiqv8XQQ0eRQRSHiXTWXL4v4/HljcjV+x9uH1OU9TCfSVkcGUcsQ3YimAhNXdTT6cKsU1MjZWzopaXa89//NKgeJ+Yn79J26dMt3EKiATb+1o5vVQx5tNDFRP9e70ZJ5h3U9KdJmBGqyvydlcoHeC8olvJfpVjIVPCmGtzUALDSwSbo5IIZg==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=amd.com; dmarc=pass action=none header.from=amd.com; dkim=pass
 header.d=amd.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=amd.com; s=selector1;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=V519WsdDsLhQcSitVWrmZPwA4RVtfnxNsUoCkWKP0qw=;
 b=ZDvMMx4KHMB4zCstn3BdOnmsi1d2wF6slYRVn3fCU/Y0PXqR/6bQN/4FvrBzz3hbScDxjAMsBppYVLRWslaHxQRhBOyINpwf/cHSCki4GAFMAlvFfzdWcS/9bz0O6LmonLdXFZmBLqtTa3S5tc+WGwK3Md5EBbn4usqAYq76FAA=
Authentication-Results: redhat.com; dkim=none (message not signed)
 header.d=none;redhat.com; dmarc=none action=none header.from=amd.com;
Received: from SN6PR12MB2767.namprd12.prod.outlook.com (2603:10b6:805:75::23)
 by SA0PR12MB4557.namprd12.prod.outlook.com (2603:10b6:806:9d::10) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4020.18; Fri, 23 Apr
 2021 15:58:54 +0000
Received: from SN6PR12MB2767.namprd12.prod.outlook.com
 ([fe80::1fb:7d59:2c24:615e]) by SN6PR12MB2767.namprd12.prod.outlook.com
 ([fe80::1fb:7d59:2c24:615e%7]) with mapi id 15.20.4065.025; Fri, 23 Apr 2021
 15:58:54 +0000
From: Ashish Kalra <Ashish.Kalra@amd.com>
To: pbonzini@redhat.com
Cc: tglx@linutronix.de, mingo@redhat.com, hpa@zytor.com,
        joro@8bytes.org, bp@suse.de, thomas.lendacky@amd.com,
        x86@kernel.org, kvm@vger.kernel.org, linux-kernel@vger.kernel.org,
        srutherford@google.com, seanjc@google.com,
        venu.busireddy@oracle.com, brijesh.singh@amd.com
Subject: [PATCH v2 2/4] mm: x86: Invoke hypercall when page encryption status
 is changed
Date: Fri, 23 Apr 2021 15:58:43 +0000
Message-Id: 
 <ff68a73e0cdaf89e56add5c8b6e110df881fede1.1619193043.git.ashish.kalra@amd.com>
X-Mailer: git-send-email 2.17.1
In-Reply-To: <cover.1619193043.git.ashish.kalra@amd.com>
References: <cover.1619193043.git.ashish.kalra@amd.com>
X-Originating-IP: [165.204.77.1]
X-ClientProxiedBy: SN1PR12CA0098.namprd12.prod.outlook.com
 (2603:10b6:802:21::33) To SN6PR12MB2767.namprd12.prod.outlook.com
 (2603:10b6:805:75::23)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ashkalra_ubuntu_server.amd.com (165.204.77.1) by
 SN1PR12CA0098.namprd12.prod.outlook.com (2603:10b6:802:21::33) with Microsoft
 SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4065.20 via Frontend
 Transport; Fri, 23 Apr 2021 15:58:53 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: 9429766f-9a2b-4379-7111-08d90670b1ce
X-MS-TrafficTypeDiagnostic: SA0PR12MB4557:
X-MS-Exchange-Transport-Forked: True
X-Microsoft-Antispam-PRVS: 
 <SA0PR12MB45570525D9541B799C85927A8E459@SA0PR12MB4557.namprd12.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:4941;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 DDmvlIgbp5/k9GkfHaobiAMLLiXFgeI7ebOFfML+Ilj7bPlMcpX4DXg6sjfXtYDhdEKjHPbxdqhhTj3UOLNjvRJuAu1waQhfJtEvFL/Ik945FEuFCPjg454VuwNoXBSnQ8t9SXTAFx1O+j616WE9Hr23B1GE6AM9S6HEC7jmVCEg+V1XVXhE+ZGN8Ny3I/Co57AAvhyzoQQ/sukLZrigoRQZjpnu8wZe1UItqsZll2/4uRjcqdsodv6S1N0fp+1eyz3QvPyaHpeOl+djJ0JD7GjEh3kL2BOGq3qd9rO54sxvjqlIeGq7UTZpDaZovvAww910ttix5box0/pCWaeBDo+prsZVSnADieJOJZlF3SrKuO3qPpe36ghzcUn+KJ9RdSf5t156wCy8+uuN3e0T2FG3iFDGhPBciEHOdtDTO6I+afAmOMK75EtHxx4SdhYbZiLe7QS/a2RcZvrOZTQfgrorRB+kcVfITq5FWFyb0vUOKGVg5vHUJPV5f1jdBhC4y4p6pTOHxx5p8co1Gou6PvorUgxDzzHAXH+ZDGhuIQUWsoDFm8drHbrz68MwL8NwZKoWeLHP5Ja8/RtTv/MYs11GbPPt/HVid/fmL0Re45KN/oqQijmi/H4YCrweguB2qq6pB2iGOtLUJhmQqgsGkQ==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR12MB2767.namprd12.prod.outlook.com;PTR:;CAT:NONE;SFS:(4636009)(366004)(136003)(396003)(39860400002)(346002)(376002)(52116002)(6916009)(7696005)(8936002)(38100700002)(38350700002)(66476007)(4326008)(2616005)(8676002)(478600001)(956004)(186003)(36756003)(5660300002)(6666004)(6486002)(316002)(26005)(7416002)(16526019)(2906002)(66556008)(66946007)(83380400001)(86362001);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 Pan1LLvsUGU+EaiRaCY4QEGuw8HSXn4Noao62CcBQmK3v02bI8HaFdy6uSas7fsU/GF5ZV42/ITyr34fE1kxAWkLi8puEfsxU8KOYM0UN39M/rX81Ai6fSJXr9b8MiiAa2TecY6zWb4TLHCAZQ03vOlNUs2My4xS3yqQAEl/QtaycT07+SwXEBsgJ5AW8Kag8zYIUwGJT+fo1Nkv8hRoM8U9L8iFTHctP2kpKcadZu2hOywqYaKsEFoQCBg/oLulLlo2uJnOQV65YAS7Lnb2KMRTfKGPHjNyltcyuM996F2Z6AxsSarKviyusgfiXk7fJ2eXSrGOFSaohnooeA16BCX04nJca0Zfh6x1w9I5OqcSWn1xOKL1JFWJ20UDTid6ZZ5l0Ney/laCyCPRa8ZYUqlrNsgMb5ID3Kx8MxdjDdKB6mNkE6UDPcHxD1Q5Z58Cf6I49nqFmJHN8vSQd7pkgsG49HpfAN/l1DiRbId97J00cyXgPv0pE34gz4KqBxGFjODavYVz58Gh8fVZUpiFI7Q4Mni+klpqqpnkoHO6qq9AveTNFMBGwOGIUU3L0yoqzlAbjVhyhbbMdumZ8DmovluG8fRDF3YGJTRPQIorD7y0WaYgNjxeIpbvWsrv8b+4PUSLWyMKu782haEKeEhwQHfkpLKSMw1BgZX0VOF9g4QSAyHIM4wyIQ4/LkUlgJWP0ZFmM6CQx95WI446gOgGDM645OXU8OA9P0sEvvOLnP/ReQaCYo/QdP9Jn73GNxgPSyoqcW8eXsEWut7vhYloFfLOzYTNH64U7Izrc4U2IFUn7WLj5PGMqbrG/cRgCaIhNToRyBQRGVi9rsPPIHze1+z674qhQEvnj0UaKOVHgE2+ETUfXxGUoIztnRHNtgEIcNEj9dyNWr91cCh5T0YiVLX7PbhgwJoYsZqcgewu/0rDqxD5lLCxPCg1o+MhR/t1z6fnb3cXe7shCjsnRxDWhgmQhUomyHFpO5OuEoFTWV1oSqXn4RZfF48ECIr1AkP0a4ugbXB4Tkryjv6gJ9XEXVdQ9FfsxdVyLCZem5YP03zyYfitakKk9qNNbko0tXbme2NAtw1UNroZOYmsLaZ0LIGnuvfQxePl2SAkA/wOjnkjdTDbwd50t32R3DqH0FcTlIXF+1QUuRnC4bEeh95132vcuZJWngFdDwTmIUlbsZj95juF0nysMnyrd7cbSr3dx7ohV9ny0NPS7BE4WX6a+N/1HCYI9YQN8Uv8rzISS5kXlENMUiyIwYcyQB82yg1LeDDUZdbIiBxDiEST+cEp/edQmhwAhnj7StbWNP428TMnZHQ4YQ9LNlYspaj+xPb3
X-OriginatorOrg: amd.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 9429766f-9a2b-4379-7111-08d90670b1ce
X-MS-Exchange-CrossTenant-AuthSource: SN6PR12MB2767.namprd12.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 23 Apr 2021 15:58:54.1454
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 3dd8961f-e488-4e60-8e11-a82d994e183d
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 0xeAgDxfW7GAHrVLyt1cTWgpPyd09BUoVDasE9AwHK8jrurN5AIZwNoVQ2Twv4bFCydBlrhAO+U7knl/ZiEOeQ==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SA0PR12MB4557
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Brijesh Singh <brijesh.singh@amd.com>

Invoke a hypercall when a memory region is changed from encrypted ->
decrypted and vice versa. Hypervisor needs to know the page encryption
status during the guest migration.

Cc: Thomas Gleixner <tglx@linutronix.de>
Cc: Ingo Molnar <mingo@redhat.com>
Cc: "H. Peter Anvin" <hpa@zytor.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Joerg Roedel <joro@8bytes.org>
Cc: Borislav Petkov <bp@suse.de>
Cc: Tom Lendacky <thomas.lendacky@amd.com>
Cc: x86@kernel.org
Cc: kvm@vger.kernel.org
Cc: linux-kernel@vger.kernel.org
Reviewed-by: Steve Rutherford <srutherford@google.com>
Reviewed-by: Venu Busireddy <venu.busireddy@oracle.com>
Signed-off-by: Brijesh Singh <brijesh.singh@amd.com>
Signed-off-by: Ashish Kalra <ashish.kalra@amd.com>
---
 arch/x86/include/asm/paravirt.h       |  6 +++
 arch/x86/include/asm/paravirt_types.h |  2 +
 arch/x86/include/asm/set_memory.h     |  2 +
 arch/x86/kernel/paravirt.c            |  1 +
 arch/x86/mm/mem_encrypt.c             | 66 +++++++++++++++++++++++----
 arch/x86/mm/pat/set_memory.c          |  7 +++
 6 files changed, 75 insertions(+), 9 deletions(-)

diff --git a/arch/x86/include/asm/paravirt.h b/arch/x86/include/asm/paravirt.h
index 4abf110e2243..c0ef13e97f5a 100644
--- a/arch/x86/include/asm/paravirt.h
+++ b/arch/x86/include/asm/paravirt.h
@@ -84,6 +84,12 @@ static inline void paravirt_arch_exit_mmap(struct mm_struct *mm)
 	PVOP_VCALL1(mmu.exit_mmap, mm);
 }
 
+static inline void notify_page_enc_status_changed(unsigned long pfn,
+						  int npages, bool enc)
+{
+	PVOP_VCALL3(mmu.notify_page_enc_status_changed, pfn, npages, enc);
+}
+
 #ifdef CONFIG_PARAVIRT_XXL
 static inline void load_sp0(unsigned long sp0)
 {
diff --git a/arch/x86/include/asm/paravirt_types.h b/arch/x86/include/asm/paravirt_types.h
index de87087d3bde..7245d08f5500 100644
--- a/arch/x86/include/asm/paravirt_types.h
+++ b/arch/x86/include/asm/paravirt_types.h
@@ -195,6 +195,8 @@ struct pv_mmu_ops {
 
 	/* Hook for intercepting the destruction of an mm_struct. */
 	void (*exit_mmap)(struct mm_struct *mm);
+	void (*notify_page_enc_status_changed)(unsigned long pfn, int npages,
+					       bool enc);
 
 #ifdef CONFIG_PARAVIRT_XXL
 	struct paravirt_callee_save read_cr2;
diff --git a/arch/x86/include/asm/set_memory.h b/arch/x86/include/asm/set_memory.h
index 4352f08bfbb5..ed9cfe062634 100644
--- a/arch/x86/include/asm/set_memory.h
+++ b/arch/x86/include/asm/set_memory.h
@@ -83,6 +83,8 @@ int set_pages_rw(struct page *page, int numpages);
 int set_direct_map_invalid_noflush(struct page *page);
 int set_direct_map_default_noflush(struct page *page);
 bool kernel_page_present(struct page *page);
+void notify_addr_enc_status_changed(unsigned long vaddr, int npages,
+				    bool enc);
 
 extern int kernel_set_to_readonly;
 
diff --git a/arch/x86/kernel/paravirt.c b/arch/x86/kernel/paravirt.c
index c60222ab8ab9..192230247ad7 100644
--- a/arch/x86/kernel/paravirt.c
+++ b/arch/x86/kernel/paravirt.c
@@ -335,6 +335,7 @@ struct paravirt_patch_template pv_ops = {
 			(void (*)(struct mmu_gather *, void *))tlb_remove_page,
 
 	.mmu.exit_mmap		= paravirt_nop,
+	.mmu.notify_page_enc_status_changed	= paravirt_nop,
 
 #ifdef CONFIG_PARAVIRT_XXL
 	.mmu.read_cr2		= __PV_IS_CALLEE_SAVE(native_read_cr2),
diff --git a/arch/x86/mm/mem_encrypt.c b/arch/x86/mm/mem_encrypt.c
index 4b01f7dbaf30..e4b94099645b 100644
--- a/arch/x86/mm/mem_encrypt.c
+++ b/arch/x86/mm/mem_encrypt.c
@@ -229,29 +229,74 @@ void __init sev_setup_arch(void)
 	swiotlb_adjust_size(size);
 }
 
-static void __init __set_clr_pte_enc(pte_t *kpte, int level, bool enc)
+static unsigned long pg_level_to_pfn(int level, pte_t *kpte, pgprot_t *ret_prot)
 {
-	pgprot_t old_prot, new_prot;
-	unsigned long pfn, pa, size;
-	pte_t new_pte;
+	unsigned long pfn = 0;
+	pgprot_t prot;
 
 	switch (level) {
 	case PG_LEVEL_4K:
 		pfn = pte_pfn(*kpte);
-		old_prot = pte_pgprot(*kpte);
+		prot = pte_pgprot(*kpte);
 		break;
 	case PG_LEVEL_2M:
 		pfn = pmd_pfn(*(pmd_t *)kpte);
-		old_prot = pmd_pgprot(*(pmd_t *)kpte);
+		prot = pmd_pgprot(*(pmd_t *)kpte);
 		break;
 	case PG_LEVEL_1G:
 		pfn = pud_pfn(*(pud_t *)kpte);
-		old_prot = pud_pgprot(*(pud_t *)kpte);
+		prot = pud_pgprot(*(pud_t *)kpte);
 		break;
 	default:
-		return;
+		return 0;
 	}
 
+	if (ret_prot)
+		*ret_prot = prot;
+
+	return pfn;
+}
+
+void notify_addr_enc_status_changed(unsigned long vaddr, int npages,
+				    bool enc)
+{
+#ifdef CONFIG_PARAVIRT
+	unsigned long sz = npages << PAGE_SHIFT;
+	unsigned long vaddr_end = vaddr + sz;
+
+	while (vaddr < vaddr_end) {
+		int psize, pmask, level;
+		unsigned long pfn;
+		pte_t *kpte;
+
+		kpte = lookup_address(vaddr, &level);
+		if (!kpte || pte_none(*kpte))
+			return;
+
+		pfn = pg_level_to_pfn(level, kpte, NULL);
+		if (!pfn)
+			continue;
+
+		psize = page_level_size(level);
+		pmask = page_level_mask(level);
+
+		notify_page_enc_status_changed(pfn, psize >> PAGE_SHIFT, enc);
+
+		vaddr = (vaddr & pmask) + psize;
+	}
+#endif
+}
+
+static void __init __set_clr_pte_enc(pte_t *kpte, int level, bool enc)
+{
+	pgprot_t old_prot, new_prot;
+	unsigned long pfn, pa, size;
+	pte_t new_pte;
+
+	pfn = pg_level_to_pfn(level, kpte, &old_prot);
+	if (!pfn)
+		return;
+
 	new_prot = old_prot;
 	if (enc)
 		pgprot_val(new_prot) |= _PAGE_ENC;
@@ -286,12 +331,13 @@ static void __init __set_clr_pte_enc(pte_t *kpte, int level, bool enc)
 static int __init early_set_memory_enc_dec(unsigned long vaddr,
 					   unsigned long size, bool enc)
 {
-	unsigned long vaddr_end, vaddr_next;
+	unsigned long vaddr_end, vaddr_next, start;
 	unsigned long psize, pmask;
 	int split_page_size_mask;
 	int level, ret;
 	pte_t *kpte;
 
+	start = vaddr;
 	vaddr_next = vaddr;
 	vaddr_end = vaddr + size;
 
@@ -346,6 +392,8 @@ static int __init early_set_memory_enc_dec(unsigned long vaddr,
 
 	ret = 0;
 
+	notify_addr_enc_status_changed(start, PAGE_ALIGN(size) >> PAGE_SHIFT,
+					enc);
 out:
 	__flush_tlb_all();
 	return ret;
diff --git a/arch/x86/mm/pat/set_memory.c b/arch/x86/mm/pat/set_memory.c
index 16f878c26667..45e65517405a 100644
--- a/arch/x86/mm/pat/set_memory.c
+++ b/arch/x86/mm/pat/set_memory.c
@@ -2012,6 +2012,13 @@ static int __set_memory_enc_dec(unsigned long addr, int numpages, bool enc)
 	 */
 	cpa_flush(&cpa, 0);
 
+	/*
+	 * Notify hypervisor that a given memory range is mapped encrypted
+	 * or decrypted. The hypervisor will use this information during the
+	 * VM migration.
+	 */
+	notify_addr_enc_status_changed(addr, numpages, enc);
+
 	return ret;
 }
 

From patchwork Fri Apr 23 15:59:01 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Ashish Kalra <Ashish.Kalra@amd.com>
X-Patchwork-Id: 12220823
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.7 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,URIBL_BLOCKED,USER_AGENT_GIT autolearn=ham autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id A803DC433ED
	for <kvm@archiver.kernel.org>; Fri, 23 Apr 2021 15:59:21 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 68C5261425
	for <kvm@archiver.kernel.org>; Fri, 23 Apr 2021 15:59:21 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S243142AbhDWP74 (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Fri, 23 Apr 2021 11:59:56 -0400
Received: from mail-co1nam11on2087.outbound.protection.outlook.com
 ([40.107.220.87]:39777
        "EHLO NAM11-CO1-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S231858AbhDWP7x (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 23 Apr 2021 11:59:53 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=Ej32xX4t606nJ5gJx3fgzJ1+1sTv8yHld3XkSHg0LHpwB7ML63UiWV7qOcIgwap5kNh1PeEIhOMSnoKWxoHnm9GjFkgfLh03Tf4U3mZd+pBOrq8Hvfkvs6Odsn92f7y6lL6HrZxJQhl0Wg4mmqZ0uGJ5Zz5JMqKj2u1Gn2ux9/F2tBxoiz1kT3pNhT6ZTihYhylMW3qhL3XhTAx+dL2eXcrdd7zy6sH9vAJrj8ND5BdaqixN/ZN1tMDyFKXoXBKwnosvUQ6pFQNCVSKcTpip71lTmve0izSCH9EXywT12YpJGaZUDNvgC6Oyg5CdH3GenMj1WwWAR40y5hBSFfiEJg==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=kP40PdU4Z8HHa6wnZVnRT/zhr1u8HNB3yWt0AcQt+JM=;
 b=JJ1sJHLx/aciHgNV17A8a2S8GHyfKVZcTC6VO4p28APlporr2nVGS+0yG7/cq4mkDq5j+1HIwPTNZ28YO1PNZEdc6P9bFc5/B0BfOqWElTPfL9quG2QJhJUh4wXiB7Ea8SAEbHNnrxN+Fl7LG/bmW8qkpq8dqsaok8XdCmTAgxHcHdv3qnVsT6punYmNCWVUGbJrSqu8lpMAiADCUmdmWlG5uMgWVJh+H+4fPoQuzI1SSmxFmHxFNIXi7+AYcEAofUIxrtIVuETBeC1brWCTZ/RcUmJuHBgnhQybmCV1iAS6S+4WXJKYBy6srK8irBCL6uRYXpCqFXgKVbn8zQPK/A==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=amd.com; dmarc=pass action=none header.from=amd.com; dkim=pass
 header.d=amd.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=amd.com; s=selector1;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=kP40PdU4Z8HHa6wnZVnRT/zhr1u8HNB3yWt0AcQt+JM=;
 b=RXa1i0ksoN14HGR7pD4spY6qmhRIDrdBuoTswiQUQKxnIrRkXfjnMZOAiofDz+XGNnajuZ+Iqn0beDPFtNcA1kzQUY8nqAtQDICpINqPO41y79TjuXdpq8R2grH7Wo4c5Ixuh65ui0Nne4VZuhNsrnkRkQNb/l11tCYOaVr97fA=
Authentication-Results: redhat.com; dkim=none (message not signed)
 header.d=none;redhat.com; dmarc=none action=none header.from=amd.com;
Received: from SN6PR12MB2767.namprd12.prod.outlook.com (2603:10b6:805:75::23)
 by SN6PR12MB2717.namprd12.prod.outlook.com (2603:10b6:805:68::29) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4042.18; Fri, 23 Apr
 2021 15:59:11 +0000
Received: from SN6PR12MB2767.namprd12.prod.outlook.com
 ([fe80::1fb:7d59:2c24:615e]) by SN6PR12MB2767.namprd12.prod.outlook.com
 ([fe80::1fb:7d59:2c24:615e%7]) with mapi id 15.20.4065.025; Fri, 23 Apr 2021
 15:59:11 +0000
From: Ashish Kalra <Ashish.Kalra@amd.com>
To: pbonzini@redhat.com
Cc: tglx@linutronix.de, mingo@redhat.com, hpa@zytor.com,
        joro@8bytes.org, bp@suse.de, thomas.lendacky@amd.com,
        x86@kernel.org, kvm@vger.kernel.org, linux-kernel@vger.kernel.org,
        srutherford@google.com, seanjc@google.com,
        venu.busireddy@oracle.com, brijesh.singh@amd.com
Subject: [PATCH v2 3/4] EFI: Introduce the new AMD Memory Encryption GUID.
Date: Fri, 23 Apr 2021 15:59:01 +0000
Message-Id: 
 <f9d22080293f24bd92684915fcee71a4974593a3.1619193043.git.ashish.kalra@amd.com>
X-Mailer: git-send-email 2.17.1
In-Reply-To: <cover.1619193043.git.ashish.kalra@amd.com>
References: <cover.1619193043.git.ashish.kalra@amd.com>
X-Originating-IP: [165.204.77.1]
X-ClientProxiedBy: SN4PR0501CA0039.namprd05.prod.outlook.com
 (2603:10b6:803:41::16) To SN6PR12MB2767.namprd12.prod.outlook.com
 (2603:10b6:805:75::23)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ashkalra_ubuntu_server.amd.com (165.204.77.1) by
 SN4PR0501CA0039.namprd05.prod.outlook.com (2603:10b6:803:41::16) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4087.15 via Frontend
 Transport; Fri, 23 Apr 2021 15:59:11 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: 9919eb28-9018-482f-62d6-08d90670bc3b
X-MS-TrafficTypeDiagnostic: SN6PR12MB2717:
X-MS-Exchange-Transport-Forked: True
X-Microsoft-Antispam-PRVS: 
 <SN6PR12MB271784571ED419117A77E28C8E459@SN6PR12MB2717.namprd12.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:4941;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 6oTDrPrNveAO8JzZIKc8xPr5m3IktJ167Od+PPJJC5NeJhArRsH17vKrOQ0ekqgK81p8i61pBxgNlRjHn+zKVHKtZINucXNJftu6RvX6cSWW317mE9h6WtzaxXW0hSM1kPih4PtRmma1bx2EX5bJ0+VsrPXSZ0ohMMD7roeFU+M0Ed/4ziKWmvfGEsHsmVBwo/Oy4nkGh+NN+RivgbHqmh78UnUZZlxj7H2+h7FnqIn3OLeJRKVLM2mTs/vqJTwqh5ZJoxved46ISL7SaBK44vgT0ibHMvPUj/dKz/XAHjQCEz0puOeF4UX3mIfrQ9YY8pHIwQQs59zTPjTo3XFhrS6cYbs6PkrhOueE3Xs8WrvkTeK2ar9kvwK9P9gGAi7d7a61PQ74Is8LVrtl5I2kyi9K8pgtdjEhX5yWISUSlaIGcuu8UzPVdmbYHRh2dmV/keyaOxfkB9jFPIcI1EY0psSBMV3o6vIl6EkQzzpv5nv8Y7Oyr4vJ7wvOEvnKv/2qjRXXWeWNG8uvq8XoM/+3k356vpwymAE0K1pXkwBg8FZqXkpCY2+MPS72ciNnuvgXODQRPpI8L9PolBFOXRIzcj47VKYTzWjBRYlcat4EliycRZVewtknAYc6ou+yImC/Ng0e7f4dZjHZEw6ODyO/0A==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR12MB2767.namprd12.prod.outlook.com;PTR:;CAT:NONE;SFS:(4636009)(366004)(136003)(39860400002)(376002)(346002)(396003)(83380400001)(66946007)(186003)(956004)(6486002)(38100700002)(5660300002)(316002)(6666004)(2906002)(66556008)(66476007)(16526019)(26005)(36756003)(2616005)(8676002)(478600001)(7416002)(86362001)(38350700002)(6916009)(7696005)(4326008)(8936002)(52116002);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 gY5US5g8gkOLNPKN91a1iIovq7bJ2dliu/Hf5D1qQEOsz4npIt2Az5wXdkmiytDk871YMqH3iIX/Wxagva6cANKdBCVLdafXkrK+kLvnBttWU7thz+c1+CA2nqMs3F9M1kcpoXSZOGKpbxPitmPhme7hFyC0cO4RCbY8WvqJAif/8UyRW2c34rx9sQiclGwpnDpHAui5dLzMc2XPReY9C4tFeWV9LpONDXaHZ/IVEJ+bWWhOvfg3PWFndTSLg2SDDP2nY8isR3K/ySfL2HP3ExWMeDPARaJoBEygdTvSHAm6f5uZEuKLzwRe2AAWBVtR7o0Ss8TdZsEFsbBzAcl0Rb9QOF4wKGaHLxBZ4wsWX8mI2spjedKVIH20eCozlMoBn4a/7vG1LdBIt3nSr0RUaPxZL6jRZuz1ch7XMXgLaC5YFFH62GyatZNTW+Kb7yYQ1tIL2vt0LtnHOvjKGZepNkd2A7vm9QOvxK8pll3FirSg7MGknAAjzrWmivCfasGGpiNl8msNrmqes7dHjE95f/kWNWFaGfuXg+EcMS3OsplOddmDkuiMbZvwqyMqOkeldIWV//S9vX38kz1C47LQiMdUVANqG4va2SkKUvHBb72RMF3aRfvb2b49/6f2RZKctN7XruRDuNh3a/RAfSeUsoBnD7SbmogQCoWCY43iEYA1TYMv6NBhGQlhke6Zykt/C5eo44P8yNTAvRRbaP32N6o0N5enV2LS0atzsWIIU1y7/B3KPuctfhVsxuKAkRFa0GhvQDU9yTZzot/gis/xfEngCPA0EKqT1JjXaml4xJnHAbGPWom5z71UICD/tHNPPyQa0LypHXX5TGwptyyPQwmrwDomhVJ7qPL0GPA1F5HNx47jdA3hlF/imTW8+ni4v4nLvlifEwEIpBNDSsNYEsn+q/vR0jmB7JrFJ/Bw8NcLvzItlTHk5b+laxNCWY9gorU+O8hf4gwGI3I+6MrlbfzxjBQ66XRR67ZVIUSrAb3rBre7lEXUfoq9URIjuqMf8x7gnQkwctHFLQ16+Ck6qrrF+lTtEvi1U/bGoiVtnZBZ0/oC+SnC4uOyngcge/qROGelZASnJr5CFz69WEYaZ42X4zE7O2fNToLUl3Vk8pd3CpdPn5xDltCRFm/dHujO9wX91qh4UBFBHyCBnRnn0xcOvK7lE7oAlNCbknrvYVXm7F3hGvyMnIg+XoC81lu4NbKIwA8+Wzwdsda2CnY2k0eKAN0vZgvXaKw13rH4S3/88pieWS9hW78Bvz+oFV7uY7vxW+AHU8JtEou7b/YBgHl6rJ3fJhYHeo2BH9qx2VYyjMMQDD895ZIG6SgJYGiV
X-OriginatorOrg: amd.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 9919eb28-9018-482f-62d6-08d90670bc3b
X-MS-Exchange-CrossTenant-AuthSource: SN6PR12MB2767.namprd12.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 23 Apr 2021 15:59:11.5367
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 3dd8961f-e488-4e60-8e11-a82d994e183d
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 qBRCBB4Iaswo/GCA8h1L4zvEU+n4LzEa9Xyl9RyaVi5AT0hGu5ACA04ERdxIY61pV6YqhECA7MT22jhs4OvBmQ==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SN6PR12MB2717
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Ashish Kalra <ashish.kalra@amd.com>

Introduce a new AMD Memory Encryption GUID which is currently
used for defining a new UEFI environment variable which indicates
UEFI/OVMF support for the SEV live migration feature. This variable
is setup when UEFI/OVMF detects host/hypervisor support for SEV
live migration and later this variable is read by the kernel using
EFI runtime services to verify if OVMF supports the live migration
feature.

Signed-off-by: Ashish Kalra <ashish.kalra@amd.com>
---
 include/linux/efi.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/linux/efi.h b/include/linux/efi.h
index 8710f5710c1d..e95c144d1d02 100644
--- a/include/linux/efi.h
+++ b/include/linux/efi.h
@@ -360,6 +360,7 @@ void efi_native_runtime_setup(void);
 
 /* OEM GUIDs */
 #define DELLEMC_EFI_RCI2_TABLE_GUID		EFI_GUID(0x2d9f28a2, 0xa886, 0x456a,  0x97, 0xa8, 0xf1, 0x1e, 0xf2, 0x4f, 0xf4, 0x55)
+#define MEM_ENCRYPT_GUID			EFI_GUID(0x0cf29b71, 0x9e51, 0x433a,  0xa3, 0xb7, 0x81, 0xf3, 0xab, 0x16, 0xb8, 0x75)
 
 typedef struct {
 	efi_guid_t guid;

From patchwork Fri Apr 23 15:59:19 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Ashish Kalra <Ashish.Kalra@amd.com>
X-Patchwork-Id: 12220825
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.7 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,URIBL_BLOCKED,USER_AGENT_GIT autolearn=ham autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id EE438C433B4
	for <kvm@archiver.kernel.org>; Fri, 23 Apr 2021 15:59:37 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id AA56961445
	for <kvm@archiver.kernel.org>; Fri, 23 Apr 2021 15:59:37 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S243167AbhDWQAN (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Fri, 23 Apr 2021 12:00:13 -0400
Received: from mail-co1nam11on2055.outbound.protection.outlook.com
 ([40.107.220.55]:48865
        "EHLO NAM11-CO1-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S243153AbhDWQAM (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 23 Apr 2021 12:00:12 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=JS2N72z4KJQC7vZNCns6tZIL5vLM5RMuibXRTNt2GqgqdyG0FrbygOS9kfAtjQ0rHhwVfOuEMH6lr0donL9VS32FIhbSS4v2k8QmZF54q6mRfBxZ6X3B110FnjDC/nrjh3PZAgE9vWpk8G/PeqkWCMDwQTSSbK7VifUcGoU1f5KP6FlSXk1QBzbPSy0LllgDRKkyE3IMFWCff6I7fxGJoWJW5hSGx9qrRBuSOFEFtS7uZeBDfcQYIAuH2RBy2FmbTtXy5uEYp4U+COR4FWDnflHuJRe5MVYhVhw1dboLAbvETQUSF2zoviNnzoCg12ZwJ5/lRbIhy6o9TUZIcHjCcw==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=Jvi4OGI1XLHC7vWpuBcyMSPQNkyfc3ZCnuTNsVbWZnE=;
 b=cBiqLxDmofi22NavNeXnIaRJ7Xc3pQH/gv5WTHvfF6Ggr2Bsj1w5TQqhQ7V9j/0ONE4uK4DIozX92h89ce7yMruxTQNC8hZOXJMnOzYZUhNBTRQKoz6uNRbmBz0YCfPqYcUfkpQGIKulAV7bjsPmYyzKUH6L5lPR7oToYGcPMHui6uk1tueeTEdoDnsiYodcP/hiW171dYYf5nFFhJRQprwFi6t0suEhtwoe4n9zgNFubrQtUtvfl65qqkQ/N2u6YMjQTHqwJynUR7oIaKnDf0keRcBpNyambjhDoj3IYB6xhxhPAmBOb4KCjayUfandVWr6VRckVXF1Bsh9MAJ34A==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=amd.com; dmarc=pass action=none header.from=amd.com; dkim=pass
 header.d=amd.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=amd.com; s=selector1;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=Jvi4OGI1XLHC7vWpuBcyMSPQNkyfc3ZCnuTNsVbWZnE=;
 b=QbDSs5vGQLuwtY9tEdS3GdxfdbZa5WZ8qlru2TQ0yvIF3ukMeUpVXceXDP0szaJUY6BgFFxenML0HemuYWsr9Fh3jwYta8FegU917RVAUykHDJ7xWy8T0fIP2dgTEy1pxN5aFwtu18VLqIDSd1M35qOKfXCQwd79hYFfWGV0iGM=
Authentication-Results: redhat.com; dkim=none (message not signed)
 header.d=none;redhat.com; dmarc=none action=none header.from=amd.com;
Received: from SN6PR12MB2767.namprd12.prod.outlook.com (2603:10b6:805:75::23)
 by SN6PR12MB2717.namprd12.prod.outlook.com (2603:10b6:805:68::29) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4042.18; Fri, 23 Apr
 2021 15:59:29 +0000
Received: from SN6PR12MB2767.namprd12.prod.outlook.com
 ([fe80::1fb:7d59:2c24:615e]) by SN6PR12MB2767.namprd12.prod.outlook.com
 ([fe80::1fb:7d59:2c24:615e%7]) with mapi id 15.20.4065.025; Fri, 23 Apr 2021
 15:59:29 +0000
From: Ashish Kalra <Ashish.Kalra@amd.com>
To: pbonzini@redhat.com
Cc: tglx@linutronix.de, mingo@redhat.com, hpa@zytor.com,
        joro@8bytes.org, bp@suse.de, thomas.lendacky@amd.com,
        x86@kernel.org, kvm@vger.kernel.org, linux-kernel@vger.kernel.org,
        srutherford@google.com, seanjc@google.com,
        venu.busireddy@oracle.com, brijesh.singh@amd.com
Subject: [PATCH v2 4/4] x86/kvm: Add guest support for detecting and enabling
 SEV Live Migration feature.
Date: Fri, 23 Apr 2021 15:59:19 +0000
Message-Id: 
 <09a5d8a6fe69d65a2418799848dd037d24b8ba95.1619193043.git.ashish.kalra@amd.com>
X-Mailer: git-send-email 2.17.1
In-Reply-To: <cover.1619193043.git.ashish.kalra@amd.com>
References: <cover.1619193043.git.ashish.kalra@amd.com>
X-Originating-IP: [165.204.77.1]
X-ClientProxiedBy: SN4PR0501CA0023.namprd05.prod.outlook.com
 (2603:10b6:803:40::36) To SN6PR12MB2767.namprd12.prod.outlook.com
 (2603:10b6:805:75::23)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ashkalra_ubuntu_server.amd.com (165.204.77.1) by
 SN4PR0501CA0023.namprd05.prod.outlook.com (2603:10b6:803:40::36) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.4087.16 via Frontend
 Transport; Fri, 23 Apr 2021 15:59:29 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: c6711535-c6d7-47e8-bd5f-08d90670c71f
X-MS-TrafficTypeDiagnostic: SN6PR12MB2717:
X-MS-Exchange-Transport-Forked: True
X-Microsoft-Antispam-PRVS: 
 <SN6PR12MB2717AA1599A884C5A50F26CD8E459@SN6PR12MB2717.namprd12.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:305;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 VkNDiOR1VyOb1dxQt2hKuqdLjwQcT52irdTCa8BtRASVg8lyeemt54dQUe2FIe5oPJ9ywPnAF6Z8/i/usKrmoKigQScxTg0sMHAEoIzzup7xZ9hsSMm5B1civMrEhVdfluYMA3SjI+xDeRpTilI6RnXYgw735JFO352a6qFwHlGEMz04WdFeoocqMjHzP9bEXPlzkOWbPE8h8zo6S0WQ9ZbOI+ohlVsjrCZBlZOmvlmY49ApmRpXAmBjzdcrVkfP5wHg2uWLYfv6QUFyiax5WlaZm1dhXK6AHudQ36zWtpl8m00RyIs+F/aBUKsOmQ2T4kTS5i+CkjI7f6YI37DB8tXkAweJ0gGEUFEhk/PyFZDs31wH3m77WitzY2s6JW0n3T4GNqqSpVc+h4kEMWan8ajZwvvda0kbJUkVEZMduGkK2W6UiU+Ai/o0QVbhWv3VOJUpI9G8MEVnw8moYFATfyrpmeOD71f2efgd15kMSTNnFSE3GKmce8qaXabijrLJWzAQXiwu4ux4gY1Y4iGu1KCcFJ5qiRVQ05vsoVPme2Kn5JJaqi6v/9JCqhk7UDK7zfHP4ZHplYL3gEdcmF7hNQxTrF1Ezz9R9GuhIQbEn85scLPluRmBSdr4qQpHV6pbj5BQZ+Kfm7quKp+xPEgoOQ==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR12MB2767.namprd12.prod.outlook.com;PTR:;CAT:NONE;SFS:(4636009)(366004)(136003)(39860400002)(376002)(346002)(396003)(83380400001)(66946007)(186003)(956004)(6486002)(38100700002)(5660300002)(316002)(6666004)(2906002)(66556008)(66476007)(16526019)(26005)(36756003)(2616005)(8676002)(478600001)(7416002)(86362001)(38350700002)(6916009)(7696005)(4326008)(8936002)(52116002);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 TZagx3Iqaf4hkPaJmMi/T8iL3MxEXfP12J4ZqKnLEvIfrcJmMuN2Al64yND646JrYiH/QQPSJeANoXu7ZeaqQ4QeW/j52TKYchLKIe61ZIfoIhLH83eHC00IDoIZifB5Zt5XTfxHSQIvX7ltcuC4uPoEwW+kJEqtpOORGRmO4riA02jhxGGdMWkmCieEtA2W+R01PzYtTAKRhTcMI8x0EeQ25nfpZ2AR2EpuF9Yw5DemY6IMhrVKQaboaEH1eWsASxkdb4SJefKvALU4OqByNuVXC3+EtvNsVjMtzg1fB5HS9oxpB/90hrBs6e/OPAAkVTypRogszIIAcnoZOwA/djBwVEXQFNbdflUrgjXzmDTJzJ/fkTSgUzQFGE+eAweeKQlBf43a4VuziRnYjYCgGPUBCfSULM7EV5m6uRen+sVHymLUNVuFwB53tFYLcFEG+kKd7yIh43Ll5+BS83QM/uq8jwDsbiuLYqpxDIraV20RvdcvcTUgSYonVR72ebJZoZpMuO3DZCmV5O9cn+3mKy4W5DcLg+wEsaPSIajoUpU/1DpKZx/iYlxt3kfrzx0GSWBcVwAAVTzDopycBdVpey20JHUJRMrPUQ/RdfE9tKUT5VWzu1LM/EjM05TPdz+zPwFrf1LrS6GqejmLQokWukY4xPiO9MqvQwO0hJ69Qyx1/FV1DV/IAgjrAG0Oid+x5Wo7QHZ8cvq7yIqQF4rUBxwpAV0J7WaQvWZpjs73c3HU/fkGYw+0xrDrIjt55vw6yA4JA9cjdukXavt+mw6vqrY/wNbUAr7N2l6hjhHCg+XixV3hdP+O7GOsaGGgfwwVWGXe7avK6NjEfY7kpqJQTAKblch043KuuMrfISUCFZBw7suljUeF7Lxv/uVjA5WSfMQkpMg7GDpYrwq1o0LUkfzBGq847HdRlAswFOeZZS+V7IsO3AynMhyA7dS2JdDmfUngIBQOcHhEQ/AJ++pBDQGl/9ZOoKocY/YTl8je4z+/B14aqsUmZ/st56tBroQ9hWuDWeqrmoswe1S3sLlb9o6Msc08wS4nJd1s2mDDxb8iPjf/Z29jZT+0e635qv6h/TrWrshOJFeXVli5ACltcubP6745IhW2LNcYNpZJyj16lQp6NZsCid0soLmQ5AE21UTJ0flT1Gfxxw1HxQjeAOIO+3r0yb4Xa5Njzshr66FLOoihZXzLdHz22wltUy1IT8IXfwvnNKWECKvDj5/eXQLKWwO3IFBg/a0OSsDIzwH7C37/3NHwaOuhmkOiykuK+LKrNcOUn3SBDXyp/XA1Tzm3Fslpg79J8+9zjs3BJqzNNWrRtjdorjH5L48uxNie
X-OriginatorOrg: amd.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 c6711535-c6d7-47e8-bd5f-08d90670c71f
X-MS-Exchange-CrossTenant-AuthSource: SN6PR12MB2767.namprd12.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 23 Apr 2021 15:59:29.7994
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 3dd8961f-e488-4e60-8e11-a82d994e183d
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 J0fM9mnF01IpuCvChVyXc+ZQBaWMyCWr6Dy2nhj3YjwqJpaETFEx9u/ktlW9id997//Zwl1VTKK3opsN5KFKtA==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SN6PR12MB2717
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Ashish Kalra <ashish.kalra@amd.com>

The guest support for detecting and enabling SEV Live migration
feature uses the following logic :

 - kvm_init_plaform() checks if its booted under the EFI

   - If not EFI,

     i) if kvm_para_has_feature(KVM_FEATURE_MIGRATION_CONTROL), issue a wrmsrl()
         to enable the SEV live migration support

   - If EFI,

     i) If kvm_para_has_feature(KVM_FEATURE_MIGRATION_CONTROL), read
        the UEFI variable which indicates OVMF support for live migration

     ii) the variable indicates live migration is supported, issue a wrmsrl() to
          enable the SEV live migration support

The EFI live migration check is done using a late_initcall() callback.

Also, ensure that _bss_decrypted section is marked as decrypted in the
shared pages list.

Also adds kexec support for SEV Live Migration.

Reset the host's shared pages list related to kernel
specific page encryption status settings before we load a
new kernel by kexec. We cannot reset the complete
shared pages list here as we need to retain the
UEFI/OVMF firmware specific settings.

The host's shared pages list is maintained for the
guest to keep track of all unencrypted guest memory regions,
therefore we need to explicitly mark all shared pages as
encrypted again before rebooting into the new guest kernel.

Signed-off-by: Ashish Kalra <ashish.kalra@amd.com>
---
 arch/x86/include/asm/mem_encrypt.h |   4 ++
 arch/x86/kernel/kvm.c              | 106 +++++++++++++++++++++++++++++
 arch/x86/mm/mem_encrypt.c          |   6 ++
 3 files changed, 116 insertions(+)

diff --git a/arch/x86/include/asm/mem_encrypt.h b/arch/x86/include/asm/mem_encrypt.h
index 31c4df123aa0..776e3378a782 100644
--- a/arch/x86/include/asm/mem_encrypt.h
+++ b/arch/x86/include/asm/mem_encrypt.h
@@ -44,6 +44,8 @@ void __init sme_enable(struct boot_params *bp);
 
 int __init early_set_memory_decrypted(unsigned long vaddr, unsigned long size);
 int __init early_set_memory_encrypted(unsigned long vaddr, unsigned long size);
+void __init early_set_mem_enc_dec_hypercall(unsigned long vaddr, int npages,
+					    bool enc);
 
 void __init mem_encrypt_free_decrypted_mem(void);
 
@@ -84,6 +86,8 @@ static inline int __init
 early_set_memory_decrypted(unsigned long vaddr, unsigned long size) { return 0; }
 static inline int __init
 early_set_memory_encrypted(unsigned long vaddr, unsigned long size) { return 0; }
+static inline void __init
+early_set_mem_enc_dec_hypercall(unsigned long vaddr, int npages, bool enc) {}
 
 static inline void mem_encrypt_free_decrypted_mem(void) { }
 
diff --git a/arch/x86/kernel/kvm.c b/arch/x86/kernel/kvm.c
index 224a7a1ed6c3..1844b200f6d4 100644
--- a/arch/x86/kernel/kvm.c
+++ b/arch/x86/kernel/kvm.c
@@ -26,6 +26,7 @@
 #include <linux/kprobes.h>
 #include <linux/nmi.h>
 #include <linux/swait.h>
+#include <linux/efi.h>
 #include <asm/timer.h>
 #include <asm/cpu.h>
 #include <asm/traps.h>
@@ -38,6 +39,7 @@
 #include <asm/cpuidle_haltpoll.h>
 #include <asm/ptrace.h>
 #include <asm/svm.h>
+#include <asm/e820/api.h>
 
 DEFINE_STATIC_KEY_FALSE(kvm_async_pf_enabled);
 
@@ -383,6 +385,8 @@ static void kvm_pv_guest_cpu_reboot(void *unused)
 	 */
 	if (kvm_para_has_feature(KVM_FEATURE_PV_EOI))
 		wrmsrl(MSR_KVM_PV_EOI_EN, 0);
+	if (kvm_para_has_feature(KVM_FEATURE_MIGRATION_CONTROL))
+		wrmsrl(MSR_KVM_MIGRATION_CONTROL, 0);
 	kvm_pv_disable_apf();
 	kvm_disable_steal_time();
 }
@@ -429,6 +433,55 @@ static inline void __set_percpu_decrypted(void *ptr, unsigned long size)
 	early_set_memory_decrypted((unsigned long) ptr, size);
 }
 
+static int __init setup_efi_kvm_sev_migration(void)
+{
+	efi_char16_t efi_sev_live_migration_enabled[] = L"SevLiveMigrationEnabled";
+	efi_guid_t efi_variable_guid = MEM_ENCRYPT_GUID;
+	efi_status_t status;
+	unsigned long size;
+	bool enabled;
+
+	if (!sev_active() ||
+	    !kvm_para_has_feature(KVM_FEATURE_MIGRATION_CONTROL))
+		return 0;
+
+	if (!efi_enabled(EFI_BOOT))
+		return 0;
+
+	if (!efi_enabled(EFI_RUNTIME_SERVICES)) {
+		pr_info("%s : EFI runtime services are not enabled\n", __func__);
+		return 0;
+	}
+
+	size = sizeof(enabled);
+
+	/* Get variable contents into buffer */
+	status = efi.get_variable(efi_sev_live_migration_enabled,
+				  &efi_variable_guid, NULL, &size, &enabled);
+
+	if (status == EFI_NOT_FOUND) {
+		pr_info("%s : EFI live migration variable not found\n", __func__);
+		return 0;
+	}
+
+	if (status != EFI_SUCCESS) {
+		pr_info("%s : EFI variable retrieval failed\n", __func__);
+		return 0;
+	}
+
+	if (enabled == 0) {
+		pr_info("%s: live migration disabled in EFI\n", __func__);
+		return 0;
+	}
+
+	pr_info("%s : live migration enabled in EFI\n", __func__);
+	wrmsrl(MSR_KVM_MIGRATION_CONTROL, KVM_PAGE_ENC_STATUS_UPTODATE);
+
+	return true;
+}
+
+late_initcall(setup_efi_kvm_sev_migration);
+
 /*
  * Iterate through all possible CPUs and map the memory region pointed
  * by apf_reason, steal_time and kvm_apic_eoi as decrypted at once.
@@ -763,8 +816,61 @@ static bool __init kvm_msi_ext_dest_id(void)
 	return kvm_para_has_feature(KVM_FEATURE_MSI_EXT_DEST_ID);
 }
 
+static void kvm_sev_hc_page_enc_status(unsigned long pfn, int npages, bool enc)
+{
+	kvm_hypercall3(KVM_HC_PAGE_ENC_STATUS, pfn << PAGE_SHIFT, npages,
+			   enc);
+}
+
 static void __init kvm_init_platform(void)
 {
+	if (sev_active() &&
+	    kvm_para_has_feature(KVM_FEATURE_MIGRATION_CONTROL)) {
+		unsigned long nr_pages;
+		int i;
+
+		pv_ops.mmu.notify_page_enc_status_changed =
+			kvm_sev_hc_page_enc_status;
+
+		/*
+		 * Reset the host's shared pages list related to kernel
+		 * specific page encryption status settings before we load a
+		 * new kernel by kexec. Reset the page encryption status
+		 * during early boot intead of just before kexec to avoid SMP
+		 * races during kvm_pv_guest_cpu_reboot().
+		 * NOTE: We cannot reset the complete shared pages list
+		 * here as we need to retain the UEFI/OVMF firmware
+		 * specific settings.
+		 */
+
+		for (i = 0; i < e820_table->nr_entries; i++) {
+			struct e820_entry *entry = &e820_table->entries[i];
+
+			if (entry->type != E820_TYPE_RAM)
+				continue;
+
+			nr_pages = DIV_ROUND_UP(entry->size, PAGE_SIZE);
+
+			kvm_hypercall3(KVM_HC_PAGE_ENC_STATUS, entry->addr,
+					   nr_pages, 1);
+		}
+
+		/*
+		 * Ensure that _bss_decrypted section is marked as decrypted in the
+		 * shared pages list.
+		 */
+		nr_pages = DIV_ROUND_UP(__end_bss_decrypted - __start_bss_decrypted,
+					PAGE_SIZE);
+		early_set_mem_enc_dec_hypercall((unsigned long)__start_bss_decrypted,
+						nr_pages, 0);
+
+		/*
+		 * If not booted using EFI, enable Live migration support.
+		 */
+		if (!efi_enabled(EFI_BOOT))
+			wrmsrl(MSR_KVM_MIGRATION_CONTROL,
+			       KVM_PAGE_ENC_STATUS_UPTODATE);
+	}
 	kvmclock_init();
 	x86_platform.apic_post_init = kvm_apic_init;
 }
diff --git a/arch/x86/mm/mem_encrypt.c b/arch/x86/mm/mem_encrypt.c
index e4b94099645b..5294e7057ce1 100644
--- a/arch/x86/mm/mem_encrypt.c
+++ b/arch/x86/mm/mem_encrypt.c
@@ -409,6 +409,12 @@ int __init early_set_memory_encrypted(unsigned long vaddr, unsigned long size)
 	return early_set_memory_enc_dec(vaddr, size, true);
 }
 
+void __init early_set_mem_enc_dec_hypercall(unsigned long vaddr, int npages,
+					bool enc)
+{
+	notify_addr_enc_status_changed(vaddr, npages, enc);
+}
+
 /*
  * SME and SEV are very similar but they are not the same, so there are
  * times that the kernel will need to distinguish between SME and SEV. The
